### PR TITLE
[vm & was] Vulnerabilities

### PR DESCRIPTION
**Pull request by harshshah-tenable**
# Description

- This PR includes the implementation of the search method for vulnerabilities.
- Implemented Actions for Vulnerability using Findings API
- Updates for documentation and unit tests is included in the changes

# usage 

VM (findings/vulnerabilities/host)

 Import Vulnerability:  `tio.v3.vm.vulnerabilities.import_vulnerability(data)`
 Search `tio.v3.vm.vulnerabilities.search(filter=('id', 'eq', '00089a45-44a5-4620-bf9f-75ebedc6cc6c'),fields=['id'],limit=50)`

WAS (findings/vulnerabilities/webapp)

Details `tio.v3.was.vulnerabilities.get_details(id)`
Search: `tio.v3.was.vulnerabilities.search(limit=200)`

### CODE REVIEW

File: tenable/io/v3/was/__init__.py [hunk 1]
```diff
@@ -65,3 +66,11 @@ def scans(self):
         :doc:`Tenable.IO v3 Scans API <scans>`
         '''
         return ScansAPI(self._api)
+
+    @property
+    def vulnerabilities(self):
+        '''
+        The interface object for the
+        :doc:`Tenable.io v3 WAS Vulnerability APIs <vulnerability>`
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/was/__init__.py]: Vulnerability  => Vulnerabilities

File: tenable/io/v3/vm/__init__.py [hunk 2]
```diff
@@ -145,3 +146,11 @@ def server(self):
         :doc:`Tenable.io v3 Server APIs <server>`
         '''
         return ServerAPI(self._api)
+
+    @property
+    def vulnerabilities(self):
+        '''
+        The interface object for the
+        :doc:`Tenable.io v3 VM Vulnerability APIs <vulnerability>`
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/__init__.py]: Vulnerability => Vulnerabilities. 
At all the places, like name of file, name of class, let's replace it with Vulnerabilities as per the API spec.

File: tenable/io/v3/vm/vulnerability/api.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    _conv_json = True
+
+    def import_vulnerability(self, data: Dict) -> Dict:
+        '''
+        The import vulnerability methods
+
+        Args:
+            data: vulnerability data the need to be imported
+            Example:
+                >>> data = {'vendor': 'tenable',
```
Comment (milanaghera-tenable) [inline tenable/io/v3/vm/vulnerability/api.py:39]: Please update the example as per the python console syntax.

File: tenable/io/v3/vm/vulnerability/api.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+                    ...  }],
+                    ...  'and': [{
+                    ...      'value': '3',
+                    ...      'operator': 'oper',
+                    ...      'property': 3
+                    ...  }]
+                    ... }
+                As the filters may change and sortable fields may change over
+                time, it's highly recommended that you look at the output of
+                the :py:meth:`tio.v3.vm.filters.audit_log_filters()`
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/vulnerability/api.py]: Please update the corresponding filter for vulnerability.

File: tenable/io/v3/vm/vulnerability/api.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+                specific page of data.
+
+            return_csv (bool, optional):
+                If set to true, it will return the CSV response or
+                iterable (based on return_resp flag). Iterator returns all
+                rows in text/csv format for each call with row headers.
+
+        Returns:
+
+            iterable:
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/vulnerability/api.py]: iterable => Iterable

File: tenable/io/v3/was/vulnerability/api.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        Args:
+            id: The unique identifier of the finding.
+
+        Returns:
+            :obj:`dict`: details of the respective finding id
+
+        Examples:
+            >>> tio.v3.was.vulnerabilities.get_details(id)
+        '''
+        return self._get(f'{id}')
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/was/vulnerability/api.py]: Let's use _details from ExploreBaseEndpoint if the functionalities are same and CSV export should also be supported for this.

File: tenable/io/v3/was/vulnerability/api.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                specific page of data.
+
+            return_csv (bool, optional):
+                If set to true, it will return the CSV response or
+                iterable (based on return_resp flag). Iterator returns all
+                rows in text/csv format for each call with row headers.
+
+        Returns:
+
+            iterable:
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/was/vulnerability/api.py]: iterable => Iterable

File: tenable/io/v3/vm/vulnerability/api.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+                    ...  'netbios_name': 'JUPITER',
+                    ...  'vulnerabilities': [{
+                    ...      'tenable_plugin_id': '97737',
+                    ...      'last_found': 1568086236,
+                    ...      'output': ('Description: The remote Windows host
+                    ...          is missing a security update.')
+                    ...  }]
+                    ... }]
+
+            product (str, Optional):
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/vulnerability/api.py]: I think Optional should be with lower case everywhere => optional.

**Review by dhvanilshah-tenable — APPROVED** [review]

**Review by parvdesai-tenable — APPROVED** [review]

File: tests/io/v3/was/vulnerabilities/test_was_vulnerabilities_api.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+    iterator = api.v3.was.vulnerabilities.search(
+        fields=fields, limit=200, sort=sort
+    )
+    assert isinstance(iterator, SearchIterator)
+
+    event_list = []
+    for event in iterator:
+        event_list.append(event)
+    assert len(event_list) == api_response['pagination']['total']
```
Comment (ngaglani-tenable) [inline tests/io/v3/was/vulnerabilities/test_was_vulnerabilities_api.py]:
```suggestion
    vulnerabilities_list = []
    for vulnerability in iterator:
        vulnerabilities_list.append(vulnerability)
    assert len(vulnerabilities_list) == api_response['pagination']['total']
```

**Review by alpesh-te — APPROVED** [review]
LGTM!

**Pull request merged**